bpf/protocol_common.h

```diff
@@ -0,0 +1,226 @@
+		int last_offset;
+	};
+	union {
+		struct iovec __ubuf_iovec;
+		struct {
+			union {
+				const struct iovec *__iov;
+				const struct kvec *kvec;
+				const struct bio_vec *bvec;
+				struct xarray *xarray;
+				void *ubuf;
+			};
+			size_t count;
+		};
+	};
+    union {
+		unsigned long nr_segs;
+		loff_t xarray_start;
+	};
+};
+
+
+static __always_inline int read_msghdr_buf(struct msghdr *msg, u8* buf, int max_len) {
+    struct iov_iter msg_iter = BPF_CORE_READ(msg, msg_iter);
+    u8 msg_iter_type = 0;
+
+    if (bpf_core_field_exists(msg_iter.iter_type)) {
+        bpf_probe_read(&msg_iter_type, sizeof(u8), &(msg_iter.iter_type));
+        bpf_dbg_printk("msg iter type exists, read value %d", msg_iter_type);
+    }
+
+    bpf_dbg_printk("iter type %d", msg_iter_type);
+
+    struct iovec *iov = NULL;
+
+    u32 l = max_len;
+    bpf_clamp_umax(l, IO_VEC_MAX_LEN);
+
+    if (bpf_core_field_exists(msg_iter.iov)) {
+        bpf_probe_read(&iov, sizeof(struct iovec *), &(msg_iter.iov));
+        bpf_dbg_printk("iov exists, read value %llx", iov);
+    } else {
+        // TODO: I wonder if there's a way to check for field existence without having to
+        // make fake structures that match the new version of the kernel code. This code
+        // here assumes the kernel iov_iter structure is the format with __iov and __ubuf_iovec.
+        struct _iov_iter _msg_iter;
+        bpf_probe_read_kernel(&_msg_iter, sizeof(struct _iov_iter), &(msg->msg_iter));
+
+        bpf_dbg_printk("new kernel, iov doesn't exist, nr_segs %d", _msg_iter.nr_segs);
+        if (msg_iter_type == 5) {
+            struct iovec vec;
+            bpf_probe_read(&vec, sizeof(struct iovec), &(_msg_iter.__ubuf_iovec));
+            bpf_dbg_printk("ubuf base %llx, &ubuf base %llx", vec.iov_base, &vec.iov_base);
+
+            bpf_probe_read(buf, l, vec.iov_base);
+            return l;
+        } else {
+            bpf_probe_read(&iov, sizeof(struct iovec *), &(_msg_iter.__iov));
+        }     
+    }
+    
+    if (!iov) {
+        return 0;
+    }
+
+    if (msg_iter_type == 6) {// Direct char buffer
+        bpf_dbg_printk("direct char buffer type=6 iov %llx", iov);
+        bpf_probe_read(buf, l, iov);
+
+        return l;
+    }
+
+    struct iovec vec;
+    bpf_probe_read(&vec, sizeof(struct iovec), iov);
+
+    bpf_dbg_printk("standard iov %llx base %llx len %d", iov, vec.iov_base, vec.iov_len);
+
+    u32 tot_len = 0;
+
+    // Loop couple of times reading the various io_vecs
+    for (int i = 0; i < 4; i++) {
+        void *p = &iov[i];
+        bpf_probe_read(&vec, sizeof(struct iovec), p);
+        // No prints in loops on 5.10
+        // bpf_printk("iov[%d]=%llx base %llx, len %d", i, p, vec.iov_base, vec.iov_len);
+        if (!vec.iov_base || !vec.iov_len) {
+            continue;
+        }
+
+        u32 remaining = IO_VEC_MAX_LEN > tot_len ? (IO_VEC_MAX_LEN - tot_len) : 0;
+        u32 iov_size = vec.iov_len < l ? vec.iov_len : l;
```

Comment:
I made one more fix here in the read loop. Sometimes the max_len size is small, e.g. 25 bytes, but the vec.iov_len is large, e.g. 65535, so the loop would terminate on line 200, not reading a thing.

I noticed this on older kernels for receive messages.